apps/content-gateway-api/src/app/maintenance/jobs/index-handling/IndexCreationJob.ts

```diff
@@ -0,0 +1,146 @@
+    processId: string;
+}
+
+/**
+ * INTERNAL OBJECTS, DO NOT USE:
+ * they are exported for testing purposes only
+ */
+
+/**io-ts codec for verifying external input*/
+const IndexSuggestionCodec = t.strict({
+    shapes: t.unknown, // We aren't using this so idc what it is
+    suggestedIndexes: t.array(
+        t.strict({
+            id: t.string,
+            impact: t.array(t.string),
+            index: t.array(t.record(t.string, t.number)),
+            namespace: t.string,
+            weight: t.number,
+        })
+    ),
+});
+
+export type IndexSuggestions = t.TypeOf<typeof IndexSuggestionCodec>;
+
+// TODO: remove hard coded process Id with automatic discovery of processes
+const makeSuggestedIndexUrl = (atlasApiInfo: AtlasApiInfo) =>
+    `https://cloud.mongodb.com/api/atlas/v1.0/groups/${atlasApiInfo.projectId}/processes/${atlasApiInfo.processId}/performanceAdvisor/suggestedIndexes`;
+
+/**
+ * Retrieves and verifies the request according to these specs
+ * https://docs.atlas.mongodb.com/reference/api/pa-suggested-indexes-get-all/
+ */
+export function _queryIndexSuggestions(
+    atlasApiInfo: AtlasApiInfo
+): TE.TaskEither<MaintenanceJobError, IndexSuggestions> {
+    const client = new DigestFetch(
+        atlasApiInfo.publicKey,
+        atlasApiInfo.privateKey,
+        {}
+    );
+    const url = makeSuggestedIndexUrl(atlasApiInfo);
+
+    // Query the endpoint for index suggestions
+    const unsafeFetch = (): Promise<Response> => client.fetch(url, {});
+    const fetch = (): TE.TaskEither<MaintenanceJobError, Response> =>
+        TE.tryCatch(
+            unsafeFetch,
+            (e) => new MaintenanceJobError(new Error(String(e)))
+        );
+
+    const parseResponse = (
+        res: Response
+    ): TE.TaskEither<MaintenanceJobError, IndexSuggestions> => {
+        return TE.tryCatch(
+            async () => {
+                const resJson = await res.json();
+                if (E.isLeft(IndexSuggestionCodec.decode(resJson)))
+                    throw new Error(
+                        `\nIndexSuggestion response failed codec decode\
+                         \nExpected IndexSuggestions, got: ${JSON.stringify(
+                             resJson
+                         )}`
+                    );
+                return resJson as IndexSuggestions;
+            },
+            (e) => new MaintenanceJobError(new Error(String(e)))
+        );
+    };
+    return pipe(fetch(), TE.chain(parseResponse));
+}
+
+export const _addIndexes =
+    (db: Db) =>
+    (
+        indexSuggestions: IndexSuggestions
+    ): TE.TaskEither<MaintenanceJobError, void> => {
+        const suggestions = indexSuggestions.suggestedIndexes;
+
+        const addIndex = (
+            indexSuggestion: typeof suggestions[number]
+        ): TE.TaskEither<MaintenanceJobError, void> => {
+            return TE.tryCatch(
+                async () => {
+                    await db
+                        .createIndex(
```

Comment:
This was something I was considering adding, and in fact there is an "impact" value returned in the query for this exact use case. More details can be found [here](https://docs.opsmanager.mongodb.com/current/performance-advisor/index-ranking/#pa-index-ranking). The score is an estimate of the total performance increase to the database from adding an index after the overhead of index specific operations are considered. Given that we want this database to scale, it seems hard to fit a hard coded number threshold to this because as the number of collections we have grows, the lower the impact of any one index will be. So for now I just let every suggestion through. I definitely think this is something we should consider though, so please let me know if you have any ideas/opinions on this.